### PR TITLE
[77] Implement CLI session exit

### LINKED ISSUE #77: Implement extra process handling in CLI to allow stop of running sessions.
As we want the Recording to be cancelled correctly by the CLI we need to create a separate subprocess which runs the actual MORR core. We then can allow the CLI to still be responsive and accept user interactions such as a stop or cancel command, as the message loop will then no longer block the CLI.

### PR DESCRIPTION
**Pull request by insightmind**
This PR closes #77 

# Summary

This PR introduces a way to cancel a running session in the CLI.
To achieve this I needed to change the message loop to use `PeekMessage` instead of `GetMessage`.
Normally there functionality is almost the same, however the PeekMessage does not fully block the queue. However please test thoroughly if the functionality remains the same.

### CODE REVIEW

File: application/Common/Shared/Utility/NativeMethods.cs [hunk 1]
```diff
@@ -22,18 +32,23 @@ public static class NativeMethods
         /// </summary>
         public static void DoWin32MessageLoop()
         {
-            int status;
-            while ((status = GetMessage(out var msg, IntPtr.Zero, 0, 0)) != 0)
+            // We set allowsQueue to true as we expect a call to this method to be purposeful.
+            shouldCancel = true;
+            while (shouldCancel)
```
Comment (sjend) [inline application/Common/Shared/Utility/NativeMethods.cs]: Could you invert the boolean? It is a bit unintuitive to set `shouldCancel` to `false` to actually cancel the iteration

File: application/Common/Shared/Utility/NativeMethods.cs [hunk 2]
```diff
@@ -22,18 +32,23 @@ public static class NativeMethods
         /// </summary>
         public static void DoWin32MessageLoop()
         {
-            int status;
-            while ((status = GetMessage(out var msg, IntPtr.Zero, 0, 0)) != 0)
+            // We set to false as we expect a call to this method to be purposeful.
+            shouldCancel = false;
+            while (!shouldCancel)
```
Comment (Gordin508) [inline application/Common/Shared/Utility/NativeMethods.cs]: The new message Loop makes the CPU usage jump um from almost 0% on the current dev branch to the core being almost fully occupied due to busy waiting
>Unlike GetMessage, the PeekMessage function does not wait for a message to be posted before returning.

As i understand, this loop will not be used on production, so it may be fine for internal testing. I'd still prefer running the blocking loop in other thread if possible.

According to the CPU profiler, the program spends about 97% of its CPU time here (was 4-6% with the old pump using GetMessage).

I also fail to see why we changed from `GetMessage` (which is blocking), to `PeekMessage` in a busy-waiting loop, which essentially makes the `DoWin32MessageLoop` function blocking still.

File: application/Common/Shared/Utility/NativeMethods.cs [hunk 3]
```diff
@@ -11,6 +11,16 @@ namespace MORR.Shared.Utility
     /// </summary>
     public static class NativeMethods
     {
+        #region Properties
+
+        /// <summary>
+        /// Defines whether a current message loop should be cancelled.
+        /// We use a bool type for this as it is atomic and should not be problematic concerning threading.
```
Comment (Gordin508) [inline application/Common/Shared/Utility/NativeMethods.cs]: I may be missing something, but i don't see why this needs to be atomatic. The only situation where this comes into place is when different threads try to start and stop the loop at roughly the same time, and if this happens, the result is unpredictable either way.

Just a question out of curiosity, doesn't need implementation change.

File: application/Core/CLI/Commands/Record/RecordCommand.cs [hunk 4]
```diff
@@ -35,10 +37,18 @@ public int Execute(RecordOptions options)
                 OutputFormatter.PrintDebug(startRecordingMessage);
                 sessionManager.StartRecording();
 
+                // We start our interactive commandline so the user
+                // can still interact with the application.
+                var commandLine = new InteractiveCommandLine();
+                commandLine.Launch();
+
                 // Run message loop required for Windows hooks
                 NativeMethods.DoWin32MessageLoop();
```
Comment (Gordin508) [inline application/Core/CLI/Commands/Record/RecordCommand.cs]: It's odd to start the message loop here, but stop it inside the `Start()` function of `InteractiveCommandLine`.

File: application/Core/CLI/Interactive/InteractiveCommandLine.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+            var userThread = new Thread(Start);
+            userThread.Start();
+        }
+
+        private void Start()
+        {
+            Console.WriteLine(startMessage);
+            while (Console.Read() != exitCommand);
+            Console.WriteLine(closingMessage);
+            NativeMethods.StopMessageLoop();
```
Comment (Gordin508) [inline application/Core/CLI/Interactive/InteractiveCommandLine.cs]: If the message loop is stopped here, it should also be started here, see my comment in `RecordCommand.cs`

File: application/Core/CLI/Interactive/InteractiveCommandLine.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+{
+    internal class InteractiveCommandLine
+    {
+        private const char exitCommand = 'x';
+        private const string startMessage = "Use 'x' and enter to stop the recording!";
+        private const string closingMessage = "Closing MORR. This may take some time!";
+
+        public void Launch()
+        {
+            var userThread = new Thread(Start);
```
Comment (Gordin508) [inline application/Core/CLI/Interactive/InteractiveCommandLine.cs]: If you start another thread anyway, why the change from the blocking `GetMessage` to the non-blocking `PeekMessage`?

File: application/Common/Shared/Utility/NativeMethods.cs (hunk 2 above)
Comment (insightmind) [reply application/Common/Shared/Utility/NativeMethods.cs]: I will try to implement this using get message. However the problem is that GetMessage blocks the thread until it receives a message, so there is no actual way to end the loop. So what I will try to do is to register a custom application message and post it, so I can manually close the loop.

File: application/Core/CLI/Commands/Record/RecordCommand.cs (hunk 4 above)
Comment (insightmind) [reply application/Core/CLI/Commands/Record/RecordCommand.cs]: If we start the message loop in the Start() method of the InteractiveCommandLine our implementation is broken. There are two major reasons for that:

1. The message loop is not spawned on the main thread which will cause it not to work correctly at all, as this is an requirement for a message loop.

2. If we start the message loop in the Start method we block the thread in which the method runs and so for user interactive cancel is no longer possible.

As calling the message loop is a necessary requirement only for CollectingModules while recording I would keep it here.

File: application/Core/CLI/Interactive/InteractiveCommandLine.cs (hunk 5 above)
Comment (insightmind) [reply application/Core/CLI/Interactive/InteractiveCommandLine.cs]: See above, we may however add an cancel handler instead which is initialized by the caller.
